File: hackerRank/data_structures/FindMergePointOfTwoLists/FindMergePointOfTwoLists.java
package data_structures.FindMergePointOfTwoLists;

import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

// https://www.hackerrank.com/challenges/find-the-merge-point-of-two-joined-linked-lists/problem
public class FindMergePointOfTwoLists {
    static class SinglyLinkedListNode {
        public int data;
        public SinglyLinkedListNode next;

        public SinglyLinkedListNode(int nodeData) {
            this.data = nodeData;
            this.next = null;
        }
    }

    static class SinglyLinkedList {
        public SinglyLinkedListNode head;
        public SinglyLinkedListNode tail;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
        }

        public void insertNode(int nodeData) {
            SinglyLinkedListNode node = new SinglyLinkedListNode(nodeData);

            if (this.head == null) {
                this.head = node;
            } else {
                this.tail.next = node;
            }

            this.tail = node;
        }
    }

    public static void printSinglyLinkedList(SinglyLinkedListNode node, String sep) {
        while (node != null) {
            System.out.print(node.data);
            node = node.next;
            if (node != null) {
                System.out.print(sep);
            }
        }
    }

    static int findMergeNode(SinglyLinkedListNode head1, SinglyLinkedListNode head2) {
        Set<SinglyLinkedListNode> nums = new HashSet<>();

        while (head1 != null) {
            nums.add(head1);
            head1 = head1.next;
        }

        while (head2 != null) {
            if (nums.contains(head2)) {
                return head2.data;
            }
            head2 = head2.next;
        }
        return 0;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tests = scanner.nextInt();

        for (int t = 0; t < tests; t++) {
            int index = scanner.nextInt();
            SinglyLinkedList ll1 = new SinglyLinkedList();
            int n1 = scanner.nextInt();

            for (int i = 0; i < n1; i++) {
                int item = scanner.nextInt();
                ll1.insertNode(item);
            }

            SinglyLinkedList ll2 = new SinglyLinkedList();
            int n2 = scanner.nextInt();

            for (int i = 0; i < n2; i++) {
                int item = scanner.nextInt();
                ll2.insertNode(item);
            }

            SinglyLinkedListNode ptr1 = ll1.head;
            SinglyLinkedListNode ptr2 = ll2.head;

            for (int i = 0; i < n1; i++) {
                if (i < index) {
                    ptr1 = ptr1.next;
                }
            }

            for (int i = 0; i < n2; i++) {
                if (i != n2-1) {
                    ptr2 = ptr2.next;
                }
            }

            ptr2.next = ptr1;

            int result = findMergeNode(ll1.head, ll2.head);
            System.out.println(result);
        }
        scanner.close();
    }
}

File: hackerRank/data_structures/FindMergePointOfTwoLists/find-merge-point-of-two-lists.py
#!/bin/python3


class SinglyLinkedListNode:
    def __init__(self, node_data):
        self.data = node_data
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_node(self, node_data):
        node = SinglyLinkedListNode(node_data)

        if not self.head:
            self.head = node
        else:
            self.tail.next = node

        self.tail = node


def print_singly_linked_list(node, sep, fptr):
    while node:
        fptr.write(str(node.data))

        node = node.next

        if node:
            fptr.write(sep)


# Complete the findMergeNode function below.
def findMergeNode(head1, head2):
    nums = set()

    while head1 is not None:
        nums.add(head1)
        head1 = head1.next

    while head2 is not None:
        if head2 in nums:
            return head2.data
        head2 = head2.next

    return 0


if __name__ == '__main__':
    tests = int(input())

    for t in range(tests):
        index = int(input())
        n1 = int(input())
        ll1 = SinglyLinkedList()

        for _ in range(n1):
            item = int(input())
            ll1.insert_node(item)

        n2 = int(input())
        ll2 = SinglyLinkedList()

        for _ in range(n2):
            item = int(input())
            ll2.insert_node(item)

        ptr1 = ll1.head
        ptr2 = ll2.head

        for i in range(n1):
            if i < index:
                ptr1 = ptr1.next

        for i in range(n2):
            if i != n2-1:
                ptr2 = ptr2.next

        ptr2.next = ptr1

        result = findMergeNode(ll1.head, ll2.head)
        print(result)
